proposals/nnnn-actors.md

```diff
@@ -65,81 +70,190 @@ extension BankAccount {
-    // state right now, and there have not been any suspension points between
-    // the place where we checked for sufficient funds and here.
-    balance = balance - amount
+#### Closures
+
+The restrictions on only allowing asynchronous access to actor-isolated declarations on `self` only work so long as we can ensure that the code in which `self` is valid is executing non-concurrently on the actor. For methods on the actor class, this is established by the rules described above: asynchronous function calls are serialized via the actor's queue, and synchronous calls are only allowed when we know that we are already executing (non-concurrently) on the actor.
+
+However, `self` can also be captured by closures. Should those closures have access to actor-isolated state on the captured `self`? Consider an example where we want to close out a bank account and distribute the balance amongst a set of accounts:
+
+```swift
+extension BankAccount {
+  func close(distributingTo accounts: [BankAccount]) async {
+    let transferAmount = balance / accounts.count
+
+    accounts.forEach { account in 
+      balance = balance - transferAmount             // is this safe?
+      Task.runDetached {
+        await account.deposit(amount: transferAmount)
+      }  
+    }
     
-    // Safe: the deposit operation is queued on the `other` actor, at which 
-    // point it will update the other account's balance.    
-    await other.deposit(amount: amount)
+    await thief.deposit(amount: balance)
   }
 }
 ```
 
+The closure is accessing (and modifying) `balance`, which is part of the `self` actor's isolated state. Once the closure is formed and passed off to a function (in this case, `Sequence.forEach`), we no longer have control over when and how the closure is executed. On the other hand, we "know" that `forEach` is a synchronous function that invokes the closure on successive elements in the sequence. It is not concurrent, and the code above would be safe.
+
+If, on the other hand, we used a hypothetical parallel for-each, we would have a data race when the closure executes concurrently on different elements:
+
+```swift
+accounts.parallelForEach { account in 
+  self.balance = self.balance - transferAmount    // DATA RACE!
+  await account.deposit(amount: transferAmount)
+}
+```
+
+In this proposal, we assume that `forEach`'s closure parameter is *non-concurrent*, meaning that it will be executed serially, and that `parallelForEach`s closure parameter is *concurrent*, meaning that it can be executed concurrently with itself or with other code. For the purposes of this proposal, an escaping closure parameter is considered to be concurrent and a non-escaping closure parameter is considered to be non-concurrent. This is an heuristic that is discussed further and refined in a separate proposal on [Preventing Data Races in the Swift Concurrency Model](https://gist.github.com/DougGregor/10db898093ce33694139d1dcd7da3397).
```

Comment:
Will we be able to annotate things as @nonconcurrent or something? 
If you remember the OSLog issue we hit -- the closure was escaping but synchronous.

Reply:
Yes, if escaping defaults to concurrent, we will want a @nonconcurrent annotation. I don't know how we can check it meaningfully, though

Reply:
Cool yeah that'd solve it :)